src/dune_rules/pkg_rules.ml

```diff
@@ -1603,84 +1615,88 @@ let source_rules (pkg : Pkg.t) =
 
 module Compiler_dependency = struct
   type t =
-    | Toolchain of Toolchain.Version.t
+    | Toolchain of Toolchain.Compiler_package.t
     (* The lockdir specifies a compiler dependency as a package. *)
     | Inside_lock_dir of Pkg.t
     (* The lockdir specifies that the system compiler toolchain will
        be used rather than one managed by dune. *)
     | System_provided
-      (* The lockdir doesn't specify a compiler dependency. This
-         doesn't mean that the compiler is not needed to build the
-         project; a dune project with no dependencies won't explicitly
-         depend on the compiler but a compiler is still needed to
-         build the project. *)
-    | No_compiler_dependency
 
   let of_context context =
     let* lock_dir = Lock_dir.get context in
     let* db = DB.get context in
     match lock_dir.ocaml with
-    | None -> Memo.return No_compiler_dependency
+    | None ->
+      (* The lockdir doesn't specify a compiler dependency. This
+         doesn't mean that the compiler is not needed to build the
+         project; a dune project with no dependencies won't explicitly
+         depend on the compiler but a compiler is still needed to
+         build the project.
+
+         In this case, default to the latest version of the base
+         compiler. *)
+      let latest_compiler_package =
+        Toolchain.Available_compiler_packages.latest_base_compiler_package
+          db.available_compiler_packages
+      in
+      Memo.return (Toolchain latest_compiler_package)
```

Comment:
Are we sure this is the right thing to do? Maybe someone uses dune to not build ocaml code? Like, for example they only use Melange.

Reply:
In that case wouldn't they still need the ocaml compiler to build melange?

That said, since we generate a lockdir when running `dune build` now, this case should never get hit as there will always be a lockdir present when building the project. I think I will remove this logic since I don't think it will make it upstream.

Reply:
We do need to make sure that we don't break the hello-world scenario in the interviews. The project initially doesn't have any dependencies, right, and in that situation we still need to be able to compile ocaml code I think. Removing this code won't break it (I'll double check) because we hard code a dependency on ocaml-base-compiler, but that's a hack that I'll remove soon, at which point we will need some other way of either making sure that all projects implicitly depend on the compiler, or making sure that the compiler is available even if it's not an explicit dependency.

Reply:
Removing this also meant that I could remove a bunch of code related to finding the latest version of the compiler.